pos_accounting/src/pool/delta/delta_data_collection/mod.rs

```diff
@@ -141,3 +149,58 @@ impl<K: Ord, T> Default for DeltaDataCollection<K, T> {
         }
     }
 }
+
+#[cfg(test)]
+pub mod test {
+    use super::*;
+
+    #[test]
+    #[rustfmt::skip]
+    fn test_combine_delta_data() {
+        use DataDelta::{Create, Delete, Modify};
+        let data = |v| Box::new(v);
+
+        assert_eq!(combine_delta_data(&Create(data(2)), Create(data(3))), Err(Error::DeltaDataCreatedMultipleTimes));
+        assert_eq!(combine_delta_data(&Create(data(2)), Modify(data(3))), Ok(DeltaMapOp::Write(DataDelta::Create(data(3)))));
+        assert_eq!(combine_delta_data(&Create(data(2)), Delete),          Ok(DeltaMapOp::Delete));
+
+        assert_eq!(combine_delta_data(&Modify(data(2)), Create(data(3))), Err(Error::DeltaDataCreatedMultipleTimes));
+        assert_eq!(combine_delta_data(&Modify(data(2)), Modify(data(3))), Ok(DeltaMapOp::Write(DataDelta::Modify(data(3)))));
+        assert_eq!(combine_delta_data(&Modify(data(2)), Delete),          Ok(DeltaMapOp::Delete));
+
+        assert_eq!(combine_delta_data(&Delete,          Create(data(3))), Ok(DeltaMapOp::Write(DataDelta::Create(data(3)))));
+        assert_eq!(combine_delta_data(&Delete,          Modify(data(3))), Err(Error::DeltaDataModifyAfterDelete));
+        assert_eq!(combine_delta_data::<i32>(&Delete,   Delete),          Err(Error::DeltaDataDeletedMultipleTimes));
+    }
+
+    #[test]
+    fn test_merge_collections() {
```

Comment:
We need much more aggressive testing. Something like proprty based testing that generates data based on a model. But this is a good start for now.

Reply:
not sure how can we achieve that with generation because these tests would hit errors very often. Maybe just manually increase test coverage? There are not that many valid combinations

Reply:
> not sure how can we achieve that with generation because these tests would hit errors very often

Usually this is done by creating a model in the tests that matches the model of the data. It's not simple but necessary for completeness in such generic models. People do this even for scripting languages. Imagine the mess! It's still done because we want to be sure that our functions follow a certain, well-defined, model, and not just work for the individual that work by coincidence.

You don't have to do this now. Just throw in a TODO. We have bigger fish to fry.